basavyr/ml-playground
# Add supervised fine-tuning capabilities for GPT2

This PR takes the [`GPT`](https://huggingface.co/openai-community/gpt2) model and applies [supervised fine-tuning](https://huggingface.co/docs/trl/en/sft_trainer) on a custom dataset. The dataset contains about 50 entries of type `{"prompt": $prompt, "response": $response}` on the topic of Nuclear Physics and Nuclear Structure. 

This is just for testing.

The implementation has a `model_train()` method and a `model_eval()` method, where the latter tries to check how the fine-tuned model behaves on new prompts following the same topic.


## Other Issues

During the development, we tried to use larger transformer models, such as [Phi3-mini](https://huggingface.co/microsoft/Phi-3.5-mini-instruct), however, this was not possible. Most probable cause might be the current infrastructure (M3 Pro MacBook with 18GB of unified memory)

The following issue occurred: `RuntimeError: MPS backend out of memory (MPS allocated: 5.05 GB, other allocations: 2.29 GB, max allowed: 6.77 GB). Tried to allocate 1024.00 MB on private pool. Use PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.0 to disable upper limit for memory allocations (may cause system failure)`

Related links:
[1](https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/9133)
[2](https://github.com/pytorch/pytorch/issues/105839)
[3](https://github.com/lllyasviel/Fooocus/issues/980)

## How to test

Make sure you have installed via `pip` all the required modules from here:
```python
import json

from datasets import load_dataset
from trl import SFTConfig, SFTTrainer
from transformers import AutoModelForCausalLM, AutoTokenizer, Trainer, TrainingArguments
from datasets import load_dataset

from transformers import pipeline
```

Run `python3 train.py`

🚀 Feel free to test other models from 🤗
